xdebug_stack.c

```diff
@@ -1288,6 +1291,22 @@ function_stack_entry *xdebug_add_stack_frame(zend_execute_data *zdata, zend_op_a
 					} else {
 						ZVAL_COPY_VALUE(&(tmp->var[tmp->varc].data), ZEND_CALL_VAR_NUM(zdata, zdata->func->op_array.last_var + zdata->func->op_array.T + i - arguments_wanted));
 					}
+					if (Z_TYPE(tmp->var[tmp->varc].data) == IS_OBJECT) {
+						/* Tell garbage collector to NOT gc the objects until the stack is removed
```

Comment:
Does this only make sense for objects, or for all variables, especially arrays?

Reply:
I'll try to reproduce this for arrays, integers etc but so far, I didn't manage to make a failing test. I would agree that other types should probably be handled too, but only if I manage to make them fail

Reply:
I've made a cross-types test (a list of vars to replace, a list of replacing values, and run the function for every combination) and I ended up still segfaulting on $var = object being replaced by another object.

So I'll investigate a bit more, and I'll check if the issue occurs for some other data types combination. I'll push the new test and PR fixes

Reply:
....... I don't get why I cannot make it fail again now (and why github tests ran OK)... :/

Reply:
That sounds good.

Can I be a pain and ask to refactor the locking itself into its own `static` function: `lock_object(function_stack_entry *fse, zval *data)` ?

The `xdebug_add_stack_frame` function is already way too big.

Reply:
Sure

Reply:
Done, I've moved the lock into a dedicated static; for the other types than objects, I couldn't make a test that fail, so I've left it as is, and added a comment to outline that "if things other than object fails, then please code them" :/